apollo-runtime/src/commonTest/kotlin/com/apollographql/apollo3/network/AuthorizationInterceptor.kt

```diff
@@ -0,0 +1,164 @@
+package com.apollographql.apollo3.network
+
+import com.apollographql.apollo3.api.http.HttpRequest
+import com.apollographql.apollo3.api.http.HttpResponse
+import com.apollographql.apollo3.mpp.currentTimeMillis
+import com.apollographql.apollo3.network.AuthorizationInterceptor.TokenProvider
+import com.apollographql.apollo3.network.http.HttpInterceptor
+import com.apollographql.apollo3.network.http.HttpInterceptorChain
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * An [HttpInterceptor] that handles authentication
+ *
+ * This is provided as is and will most likely need to be adapted to different backend requirements.
+ *
+ * There is surprising amount of details that can differ between different implementation.
+ *
+ * This [AuthorizationInterceptor] assumes the tokens form a chain where each new token
+ * is computed based on the previous token value and invalidates all previous ones.
+ *
+ * @param tokenProvider a [TokenProvider] that gets tokens from the preferences or from the network
+ * @param maxSize The maximum number of links to keep. This is theoretically needed in case some requests are
+ * **very** slow and the token has been refreshed (by other concurrent requests) multiple times
+ * when the initial request receives the 401.
+ *
+ * In practice, this is very unlikely to happen and a max size of 1 should be enough for most
+ * scenarios
+ */
+class AuthorizationInterceptor(private val tokenProvider: TokenProvider, private val maxSize: Int = 1) : HttpInterceptor {
+  private val mutex = Mutex()
+
+  class Token(val value: String, val expiresEpochSecond: Long)
+
+  interface TokenProvider {
+    /**
+     * Load the token from preferences the first time a token is required.
+     *
+     * This function is never called concurrently. Implementation do not need to lock.
+     *
+     * @return the token from the preferences or null if no token was ever generated
+     */
+    suspend fun loadToken(): Token?
+
+    /**
+     * Refreshes an existing token. This is called when a token is expired or a 401
+     * error is received.
+     *
+     * This function is never called concurrently. Implementation do not need to lock.
+     *
+     * Any exception thrown will bubble up to the caller
+     *
+     * @param oldToken the previous token or null if there was none
+     * @return a new token from the oldToken
+     */
+    suspend fun refreshToken(oldToken: String?): Token
+  }
+
+  /**
+   * A token link in the chain
+   */
+  class TokenLink(
```

Comment:
I'm not quite sure I follow why we need to keep a chain of tokens, since only the most recent one is supposed to be valid?

Reply:
That's because of this test: https://github.com/apollographql/apollo-kotlin/pull/4068/files#diff-54b37c733642763f10a821c0cec37d9562c5c4ec48f24c1e499100b13e9515d5R162

In theory, you can have a very long request blocked on a very old token. If other concurent requests trigger mutliple refreshes, you need to "follow the link" to get the latest one or else you're taking the risk of generating an extra token.

It's all very theoretical but the "maxSize == 1" is "just" a special case

Reply:
I see! Thanks :)